plugins/ssr/ssrcommon.ml

```diff
@@ -741,12 +741,26 @@ let mkRefl t c gl =
   let (sigma, refl) = EConstr.fresh_global (pf_env gl) sigma Coqlib.((build_coq_eq_data()).refl) in
   EConstr.mkApp (refl, [|t; c|]), { gl with sigma }
 
+let safe_apply_type newcl args =
+  Proofview.Goal.enter begin fun gl ->
+    let env = Proofview.Goal.env gl in
+    let store = Proofview.Goal.extra gl in
+    Refine.refine ~typecheck:true begin fun sigma ->
+      let newcl = Reductionops.nf_betaiota sigma newcl (* As in former Logic.refine *) in
+      let (sigma, ev) =
+        Evarutil.new_evar env sigma ~principal:true ~store newcl in
+      (sigma, EConstr.applist (ev, args))
+    end
+  end
+
```

Comment:
Instead of using a full-blown check in refine, I'd rather just check that `newcl` is well-typed in the current environment. That's, as far as I can tell, the only potential source of unsoundness of the tactic, and it is likely to be more efficient.

Reply:
Wrt the current bug, I agree the ill-typed thing is the new conclusion (`newcl`).
In general nobody prevents `args` to be crap as well.

I think that I should both check `newcl` and `args` i.e. all inputs, and this is what refine does (it also checks the type of `ev args` matches the old conclusion, that again is important for the tactic to be safe).